worker/uniter/uniter.go

```diff
@@ -608,6 +617,21 @@ func (u *Uniter) charmState() (bool, *corecharm.URL, int, error) {
 		return canApplyCharmProfile, charmURL, charmModifiedVersion, errors.Trace(err)
 	}
 
+	charmDirExists, err := u.charmDirExists()
+	if err != nil {
+		return canApplyCharmProfile, charmURL, charmModifiedVersion, errors.Trace(err)
+	}
+	if !charmDirExists {
+		u.logger.Tracef("start to re-download charm because charm dir has gone which is usually caused by operator pod re-scheduling")
+		op, err := u.operationFactory.NewResolvedUpgrade(charmURL)
```

Comment:
Not sure we need a ResoledUpgrade - maybe just Upgrade will do.
With resolved = true, the logging will be "continue upgrade ..." and is intended to recover from a previously failed upgrade. That's not quite accurate here.